Sources/NIOExtras/TaggedRequestResponseHandler.swift

```diff
@@ -0,0 +1,130 @@
+    ///
+    /// - parameters:
+    ///    - initialBufferCapacity: `RequestResponseHandler` saves the promises for all outstanding responses in a
+    ///          buffer. `initialBufferCapacity` is the initial capacity for this buffer. You usually do not need to set
+    ///          this parameter unless you intend to pipeline very deeply and don't want the buffer to resize.
+    public init(initialBufferCapacity: Int = 4) {
+        self.promiseBuffer = [:]
+        self.promiseBuffer.reserveCapacity(initialBufferCapacity)
+    }
+
+    public func channelInactive(context: ChannelHandlerContext) {
+        switch self.state {
+        case .error:
+            // We failed any outstanding promises when we entered the error state and will fail any
+            // new promises in write.
+            assert(self.promiseBuffer.count == 0)
+        case .operational:
+            let promiseBuffer = self.promiseBuffer
+            self.promiseBuffer.removeAll()
+            promiseBuffer.forEach { promise in
+                promise.value.fail(NIOExtrasErrors.ClosedBeforeReceivingResponse())
+            }
+        }
+        context.fireChannelInactive()
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        guard self.state.isOperational else {
+            // we're in an error state, ignore further responses
+            assert(self.promiseBuffer.count == 0)
+            return
+        }
+
+        let response = self.unwrapInboundIn(data)
+        if let promise = self.promiseBuffer.removeValue(forKey: response.requestID) {
+            promise.succeed(response)
+        } else {
+            context.fireErrorCaught(NIOExtrasErrors.ResponseForInvalidRequest<Response>(id: response.requestID))
+        }
+    }
+
+    public func errorCaught(context: ChannelHandlerContext, error: Error) {
+        guard self.state.isOperational else {
+            assert(self.promiseBuffer.count == 0)
+            return
+        }
+        self.state = .error(error)
+        let promiseBuffer = self.promiseBuffer
+        self.promiseBuffer.removeAll()
+        context.close(promise: nil)
+        promiseBuffer.forEach {
+            $0.value.fail(error)
+        }
+    }
+
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        let (request, responsePromise) = self.unwrapOutboundIn(data)
+        switch self.state {
+        case .error(let error):
+            assert(self.promiseBuffer.count == 0)
+            responsePromise.fail(error)
+            promise?.fail(error)
+        case .operational:
+            self.promiseBuffer[request.requestID] = responsePromise
+            context.write(self.wrapOutboundOut(request), promise: promise)
+        }
+    }
+}
+
+extension NIOExtrasErrors {
+    public struct ResponseForInvalidRequest<Response: NIORequestIdentifiable>: NIOExtrasError, Equatable {
+        var id: Response.RequestID
```

Comment:
This should be `public` and we'll also need an `init`

Reply:
thanks, done!